File: backend/app/schemas/log.py
"""
日志相关 Schema
"""
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict


class OperationLogInfo(BaseModel):
    """操作日志信息"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: Optional[int] = None
    username: Optional[str] = None
    module: str
    action: str
    target_type: Optional[str] = None
    target_id: Optional[int] = None
    target_name: Optional[str] = None
    old_value: Optional[str] = None
    new_value: Optional[str] = None
    ip_address: Optional[str] = None
    response_code: Optional[int] = None
    response_time_ms: Optional[int] = None
    remark: Optional[str] = None
    created_at: Optional[datetime] = None


class SystemLogInfo(BaseModel):
    """系统日志信息"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    log_level: str
    module: Optional[str] = None
    message: str
    exception: Optional[str] = None
    stack_trace: Optional[str] = None
    created_at: Optional[datetime] = None


class CommunicationLogInfo(BaseModel):
    """通讯日志信息"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    device_id: Optional[int] = None
    comm_type: Optional[str] = None
    protocol: Optional[str] = None
    request_data: Optional[str] = None
    response_data: Optional[str] = None
    status: Optional[str] = None
    error_message: Optional[str] = None
    duration_ms: Optional[int] = None
    created_at: Optional[datetime] = None

File: frontend/src/api/modules/log.ts
/**
 * 日志查询 API
 */
import request from '@/utils/request'
import type { PageParams, PageResponse, TimeRangeParams, ExportParams } from './types'

export interface OperationLog {
  id: number
  user_id: number
  username: string
  module: string
  action: string
  target_type: string
  target_id: number
  target_name: string
  old_value: string
  new_value: string
  ip_address: string
  response_code: number
  response_time_ms: number
  remark: string
  created_at: string
}

export interface SystemLog {
  id: number
  log_level: 'debug' | 'info' | 'warning' | 'error' | 'critical'
  module: string
  message: string
  exception: string
  stack_trace: string
  created_at: string
}

export interface CommunicationLog {
  id: number
  device_id: number
  device_name: string
  comm_type: 'request' | 'response' | 'error'
  protocol: string
  request_data: string
  response_data: string
  status: 'success' | 'failed' | 'timeout'
  error_message: string
  duration_ms: number
  created_at: string
}

export interface LogStatistics {
  total: number
  by_level: Record<string, number>
  by_module: Record<string, number>
  by_user: { user_id: number; username: string; count: number }[]
  trend: { date: string; count: number }[]
}

/**
 * 获取操作日志
 */
export function getOperationLogs(params?: PageParams & TimeRangeParams & {
  user_id?: number
  module?: string
  action?: string
  keyword?: string
}): Promise<PageResponse<OperationLog>> {
  return request.get('/v1/logs/operations', { params })
}

/**
 * 获取系统日志
 */
export function getSystemLogs(params?: PageParams & TimeRangeParams & {
  log_level?: string
  module?: string
  keyword?: string
}): Promise<PageResponse<SystemLog>> {
  return request.get('/v1/logs/systems', { params })
}

/**
 * 获取通讯日志
 */
export function getCommunicationLogs(params?: PageParams & TimeRangeParams & {
  device_id?: number
  protocol?: string
  status?: string
}): Promise<PageResponse<CommunicationLog>> {
  return request.get('/v1/logs/communications', { params })
}

/**
 * 导出日志
 */
export function exportLogs(params: ExportParams & {
  log_type: 'operation' | 'system' | 'communication'
}): Promise<Blob> {
  return request.get('/v1/logs/export', {
    params,
    responseType: 'blob'
  })
}

/**
 * 获取日志统计
 */
export function getLogStatistics(params?: TimeRangeParams & {
  log_type?: 'operation' | 'system' | 'communication'
}): Promise<LogStatistics> {
  return request.get('/v1/logs/statistics', { params })
}
